### ISSUE TITLE
manifest-generate-paths monorepo support for Bitbucket

### ISSUE THREAD

**Issue by jackevans43** [argoproj/argo-cd#18968]
# Summary

Support `argocd.argoproj.io/manifest-generate-paths` [Manifest Paths Annotation](https://argo-cd.readthedocs.io/en/stable/operator-manual/high_availability/#manifest-paths-annotation) for Bitbucket to prevent spurious refreshes when multiple Applications use the same git monorepo stored in Bitbucket.

# Motivation

See #9472, where this ticket was spawned from.

# Proposal

Add support to the bitbucket webhook receiver to implement support the `argocd.argoproj.io/manifest-generate-paths` annotation.

**Comment by nitishfy** [comment]
I'd like to take this up.

**Comment by nitishfy** [comment]
Hi, I've been playing with the annotation from last week and tried to compare the changes by triggering webhook for github monorepo setup as well as bitbucket. On Bitbucket, inspite of adding the webhook, there is a manual refresh that need to happen (whether you specify the annotation or not). Can anyone provide some context on what should be the expected behaviour on bitbucket?

**Comment by nitishfy** [comment]
Here are few things I've discovered related to fixing this issue:

The [changed files are always 0](https://github.com/argoproj/argo-cd/blob/master/util/webhook/webhook.go#L187) in case of Bitbucket push event. [This is the reason](https://github.com/argoproj/argo-cd/blob/master/util/app/path/path.go#L119) why the annotation couldn’t be supported because Bitbucket doesn’t support changedFiles. Hence, even if the annotation is applied, the refresh will occur for all the applications. [This was added](https://github.com/argoproj/argo-cd/issues/18289) to trigger refresh in case of Webhook trigger.

In other words, the assumption that we have specified is a bit wrong and prevents us from supporting the manifest path annotation for Bitbucket as well as Azure Devops. One way to solve this is either by calling the Bitbucket API but that would require us to put extra efforts for authentication. The other simple way (a bit tricky) is to have bool value and check if the event is specifically happening for Bitbucket or Azure Devops, and then follow a completely different logic.

**Comment by leoluz** [comment]
@jackevans43 Please check the answers provided by @nitishfy above and reopen if the issue persists.

**Comment by anandf** [comment]
Students from IIIT, Bangalore are working on this, appraently we are trying to use the BitBucket diff/diffstat APIs to deduce the changed files from the webhook.

**Comment by anandf** [comment]
There is a ticket in BitBucket too raised for supporting changed files in Webhooks and why the support was removed.
https://jira.atlassian.com/browse/BCLOUD-11688

BB team is suggesting to use the diff api to get the list of changed files.
https://developer.atlassian.com/cloud/bitbucket/rest/api-group-commits/#api-repositories-workspace-repo-slug-diff-spec-get

**Comment by anandf** [comment]
The challenge is to make a `diff` or `diffstat` API call to bitbucket if the repo is private. It would need an auth token for the same and passing the auth token as a secret may not be straight forward.
